include/stl2/view/drop.hpp

```diff
@@ -47,36 +49,62 @@ STL2_OPEN_NAMESPACE {
 
 			constexpr R base() const { return base_; }
 
-			constexpr auto begin() requires !SimpleView<R> { return begin_impl(*this); }
-			constexpr auto begin() const requires Range<R const>
+			constexpr auto begin()
+			requires !(SimpleView<R> && RandomAccessRange<R>)
+			{ return begin_impl(*this); }
+			constexpr auto begin() const requires Range<R const> && RandomAccessRange<R const>
 			{ return begin_impl(*this); }
 
-			constexpr auto end() requires !SimpleView<R> { return end_impl(*this); }
-			constexpr auto end() const requires Range<R const>
+			constexpr auto end()
+			requires !(SimpleView<R> && RandomAccessRange<R>)
+			{ return end_impl(*this); }
+			constexpr auto end() const requires Range<R const> && RandomAccessRange<R const>
 			{ return end_impl(*this); }
 
 			constexpr auto size() requires !SimpleView<R> && SizedRange<R> { return size_impl(*this); }
 			constexpr auto size() const requires SizedRange<R const> { return size_impl(*this); }
 		private:
 			R base_;
 			D count_;
-			mutable detail::non_propagating_cache<iterator_t<R>> begin_;
+
+			constexpr auto& cached_begin() noexcept
+			{
+				return static_cast<typename drop_view::non_propagating_cache&>(*this);
+			}
 
 			template <class X>
 			static constexpr auto begin_impl(X& x)
 			{
-				if (!x.begin_) {
+				[[maybe_unused]] auto checked_begin = [&x]{
+					D const dist = __stl2::distance(x.base_);
+					STL2_EXPECT(x.count_ >= 0);
+					return __stl2::next(__stl2::begin(x.base_), dist < x.count_ ? dist : x.count_);
+				};
+
+				[[maybe_unused]] auto unchecked_begin = [&x]{
+					return __stl2::next(__stl2::begin(x.base_), x.count_);
+				};
+
+				if constexpr (RandomAccessRange<__maybe_const<is_const_v<X>, R>>) {
 					if constexpr (SizedRange<__maybe_const<is_const_v<X>, R>>) {
-						D const dist = __stl2::distance(x.base_);
-						STL2_EXPECT(x.count_ >= 0);
-						x.begin_ = __stl2::next(__stl2::begin(x.base_), dist < x.count_ ? dist : x.count_);
+						return checked_begin();
 					}
 					else {
-						x.begin_ = __stl2::next(__stl2::begin(x.base_), x.count_);
+						return unchecked_begin();
 					}
 				}
+				else {
```

Comment:
Would binding a reference to `x.cached_begin()` here make the rest of this more readable?